### PR TITLE
Bind error handler to instance (v3)

### PR DESCRIPTION
**Pull request by greguz**
#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: test/request-error.test.js [hunk 1]
```diff
@@ -73,3 +73,84 @@ test('default 400 on request error with custom error handler', t => {
...
+    t.strictEqual(res.statusCode, 400)
+    t.strictEqual(res.headers['content-type'], 'application/json; charset=utf-8')
+    t.deepEqual(res.json(), {
+      error: 'Bad Request',
+      message: 'Simulated',
+      statusCode: 400
+    })
+    t.strictEqual(fastify.hello, undefined)
+  })
+})
```
Comment (mcollina) [inline test/request-error.test.js:156]: Can you use async-await here? I think we would gain in readability.

Comment (greguz) [reply test/request-error.test.js:151]: You're right.

**Comment by greguz** [comment]
@Ethan-Arrowood I couldn't see any `(this: FastifyInstance, ...)` inside the `next` branch, so I just skipped the types update because I wasn't sure why those types were missing. For instance, inside the `RouteHandlerMethod` type, I was expecting a `this: FastifyInstance`. Is It normal?

**Comment by Ethan-Arrowood** [comment]
I was hoping we wouldn't have to explicitly type this, and that it would be automatic, but I've come to learn that requires things being typed as `functions` rather than interfaces. We recently merged something that added back a `this: FastifyInstance` type. `this` typing is one of hardest challenges when writing type definitions so I unfortunately don't have a best path recommendation here other than to test and see what is necessary.

**Comment by jsumners** [comment]
> Again, its a very hard problem with no easy answer

Oh, I have an easy answer.... 😉

**Comment by mcollina** [comment]
I propose we open an issue to update types.

**Comment by Ethan-Arrowood** [comment]
@jsumners rewrite everything with Flow right? 😋 

Opening an issue and linking to this pr now for tracking 😁

**Comment by jsumners** [comment]
> @jsumners rewrite everything with Flow right? 😋

giphy

**Pull request merged**